src/api.ts

```diff
@@ -0,0 +1,248 @@
+
+class APIConnection
+{
+    sock:Socket = null;
+    good: boolean = false;
+    reconnect_interval: number = null;
+    name:string = "API";
+    port:number;
+    callback:CallableFunction;
+    onCloseCallback: (hadError: any) => void;
+    onExpectedCloseCallback: (hadError: any) => void;
+    onErrorCallback: (hadError: any) => void;
+    onDataCallback: (hadError: any) => void;
+    buffer:string = "";
+
+    constructor(name:string, port:number, callback:CallableFunction) {
+        this.name = name;
+        this.port = port;
+        this.callback = callback;
+
+        var self = this;
+        this.onCloseCallback = function(hadError){self.onClose(hadError);};
+        this.onExpectedCloseCallback = function(hadError){self.onExpectedClose(hadError);}
+        this.onErrorCallback = function(hadError){self.onError(hadError);};
+        this.onDataCallback = function(hadError){self.onData(hadError);};
+
+        this.connect();
+    }
+
+    send(data:Object) {
+        try {
+            let r: string = JSON.stringify(data) + '\0';
+            info(this.name+' sending:', r.length, r);
+            this.end(r);
+        } catch(e) {
+            printException('error sending '+this.name, e);
+        }
+    }
+
+    destroy() {
+        if(!this.sock) return;
+
+        info(this.name+'.destroy()');
+        try {
+            this.sock.off('close', this.onCloseCallback);
+        } catch(e) {
+            printException('error closing event handlers in '+this.name+'.close', e);
+        }
+
+        try {
+            this.sock.off('close', this.onExpectedCloseCallback);
+        } catch(e) {
+            printException('error closing event handlers in '+this.name+'.close', e);
+        }
+
+        try {
+            this.sock.destroy(null);
+        } catch(e) {
+            printException('error destroying old '+this.name+' connection ', e);
+        }
+
+        this.sock = null;
+    }
+
+    private onError(hadError: boolean) {
+        info('error in '+this.name+' connection, will keep trying...');
+        if(this.good) {
+            park.postMessage(
+                {type: 'blank', text: 'error in '+this.name+' connection, will keep trying...'} as ParkMessageDesc
+            );
+        }
+        if(settings.rando_archipelago){
```

Comment:
we need to move this code out of this file, just provide an onError callback function to the constructor